llext: export symbols by groups

JordanYates opened the pull request:
Add the concept of "groups" to exported symbols, that allows including or excluding those groups from the exported symbol list through Kconfig.

Alternative to #98890

JordanYates committed:
llext: symbol: group exported symbols

Group exported symbols together by subsystem/library. Groups can then
be enabled or disabled together through the existance of a Kconfig
symbol named `CONFIG_LLEXT_EXPORT_SYMBOL_GROUP_{GROUP_NAME}`.

All symbols declared with the standard `EXPORT_SYMBOL` and
`EXPORT_SYMBOL_NAMED` macros fall into the `DEFAULT` group.

Signed-off-by: Jordan Yates <jordan@embeint.com>

DEFAULT

DEFAULT

JordanYates committed:
llext: symbol: custom groups

Allow symbols to be placed in custom groups, which can be included or
excluded from the build with a common Kconfig symbol.

Signed-off-by: Jordan Yates <jordan@embeint.com>

JordanYates committed:
llext: optional syscall symbol exports

Make the export of syscall symbols optional.

Signed-off-by: Jordan Yates <jordan@embeint.com>

JordanYates committed:
llext: optional libc symbol exports

Make the export of standard library symbols optional.

Signed-off-by: Jordan Yates <jordan@embeint.com>

JordanYates committed:
device: transition device exports to group symbols

Transition the Zephyr device exports to the new group symbol
abstraction.

Signed-off-by: Jordan Yates <jordan@embeint.com>

pillo79 commented on subsys/llext/llext_export.c:
These should be in a TOOLCHAIN group maybe?

pillo79 replied:
I don't use that toolchain, so I don't really know. TBH also leaving them in the common group has merit - they seem to be required by that toolchain. Ignore this comment :+1:

JordanYates committed:
llext: add helper symbol for default groups

Add a symbol that controls whether the default set of LLEXT groups are
exported from the application.

Signed-off-by: Jordan Yates <jordan@embeint.com>

pillo79 reviewed (commented):
Love this idea, backwards compatible and modular/extensible in the future. Will also use the same in the Arduino module, so +2 from me! 🙂
A few suggestions below.

teburd reviewed (dismissed):
Thank you for this, love the idea. The "DEFAULT" naming is a bit odd though I tend to agree with others, not sure I have a better answer though so we can move on.

pillo79 commented on subsys/llext/Kconfig:
```diff
@@ -10,6 +10,36 @@ menuconfig LLEXT
 
 if LLEXT
 
+config LLEXT_EXPORT_DEFAULT
```
Personally I have no issue with the "DEFAULT" group name; I was just confused by the "default groups" config looking like _the default group_ itself. 
If you fancy changing `SYMBOL_GROUP_DEFAULT`, to make the difference even more clear, I would probably vote for `SYMBOL_GROUP_NONE`. The plain truth :grin:

mathieuchopstm reviewed (commented):
Minor details but LGTM overall.

Maybe the documentation should be extended? 🙂

Especially since my understanding is that *by default groups are not exported*, so an application trying to create custom groups is responsible for creating the corresponding Kconfig symbols beforehand or it'll have a nasty surprise.

For the default group... how about `UNASSIGNED`? (thinking about the documentation: "if a symbol is not explicitly assigned to a group, it will be placed in the `UNASSIGNED` group") I do agree `DEFAULT` is kinda bad

mathieuchopstm commented on include/zephyr/llext/symbol.h:
```diff
@@ -155,33 +156,86 @@ struct llext_symtable {
+/* LLEXT application: Export if relevant group is enabled */
+#define Z_EXPORT_SYMBOL_NAMED_IN_GROUP(group, sym_ident, sym_name)		\
+	IF_ENABLED(CONFIG_LLEXT_EXPORT_SYMBOL_GROUP_ ## group,			\
+		(Z_EXPORT_SYMBOL_NAMED(sym_ident, sym_name)))
```
Actually, one remark I made in [this comment](https://github.com/zephyrproject-rtos/zephyr/pull/98890#issuecomment-3492343266):
```suggestion
	COND_CODE_1(CONFIG_LLEXT_EXPORT_SYMBOL_GROUP_ ## group,			\
		(Z_EXPORT_SYMBOL_NAMED(sym_ident, sym_name)),                   \
		(z_export_symbol_in_discarded_symbol_table(...))
```

But I won't block for this since it's more of a QoL feature - we can create an improvement issue to track this when PR is merged, and I'd be fine with implementing it myself (gives me a good excuse to get my hands dirty in LLEXT again... reminds me there's a cleanup I've been meaning to do since long but haven't gotten around to 🙂).

JordanYates committed:
doc: llext: document symbol groups

Add documentation for LLEXT symbol groups and the default groups created
by Zephyr.

Signed-off-by: Jordan Yates <jordan@embeint.com>

mathieuchopstm commented on doc/services/llext/config.rst:
```diff
@@ -126,6 +126,44 @@ LLEXT subsystem to optimize memory footprint in this case.
+
+   config LLEXT_EXPORT_SYMBOL_GROUP_{GROUP_NAME}
+      bool "Export all symbols from the {GROUP_NAME} group"
+
```
I'd add a
```suggestion
.. warning::

  The build system does not detect when a group is used in C code but has no
  corresponding Kconfig symbol. All symbols part of such groups will be silently
  dropped from the export table, as if the Kconfig symbol *was* defined but had
  been assigned value `n`.
```
to really get the point across in big scary red box 🙂

mathieuchopstm commented on doc/services/llext/config.rst:
```diff
@@ -126,6 +126,44 @@ LLEXT subsystem to optimize memory footprint in this case.
+  ``UNASSIGNED``, :kconfig:option:`CONFIG_LLEXT_EXPORT_SYMBOL_GROUP_UNASSIGNED`, Symbols without an explicit group
+  ``SYSCALL``, :kconfig:option:`CONFIG_LLEXT_EXPORT_SYMBOL_GROUP_SYSCALL`, Zephyr kernel system calls
+  ``LIBC``, :kconfig:option:`CONFIG_LLEXT_EXPORT_SYMBOL_GROUP_LIBC`, C standard library functions (:c:func:`memcpy` etc)
+  ``DEVICE``, :kconfig:option:`CONFIG_LLEXT_EXPORT_SYMBOL_GROUP_DEVICE`, Devicetree devices
```
Super pedantic nit: you can define devices that do not correspond to Devicetree nodes 🙂 (i.e., `DEVICE_DEFINE(...)`)

I'm fine with the current wording though; feel free to keep as is, or update it for more accuracy (e.g., `Zephyr device objects`).

mathieuchopstm reviewed (commented):
LGTM in current state; some suggestions:
